### ISSUE TITLE
go plugin: the usage of `plugin` prevent `dlfcn` from working

### ISSUE THREAD

**Issue by Sora233** [golang/go#50500]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.17.2 linux/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/mnt/e/coding/go/pkg/mod"
GOOS="linux"
GOPATH="/mnt/e/coding/go"
GOPRIVATE=""
GOPROXY="https://goproxy.io,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.17.2"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/mnt/e/coding/go/src/github.com/Sora233/test-plugin/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1550186777=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?
I have a project `test-plugin` with code blow:
```go
package main

/*
#cgo LDFLAGS: -ldl -Wl,--unresolved-symbols=ignore-in-object-files

#include <dlfcn.h>

void say_hello();

*/
import "C"

import (
	// ! NOTE HERE !
	// _ "plugin"
	"unsafe"
)

func main() {
	handle := C.dlopen(C.CString("libplugin.so"), C.RTLD_LAZY|C.RTLD_GLOBAL)
	if handle == C.NULL {
		panic("dlopen got null")
	}
	C.dlerror()
	C.dlsym(handle, C.CString("say_hello"))
	if unsafe.Pointer(C.dlerror()) != C.NULL {
		panic("dlsym error")
	}
	C.say_hello()
}
```

At first, I use dlopen to load my other dynamic library with some flags to avoid compile-time dependency.

At this time all is good, and I can get a go-runtime error.
```
panic: dlopen got null

goroutine 1 [running]:
main.main()
        /mnt/e/coding/go/src/github.com/Sora233/test-plugin/main.go:21 +0x108
```

But after I import the `plugin` package I can not get an go-runtime error.

I got `./test-plugin: symbol lookup error: ./test-plugin: undefined symbol: say_hello` instead.

And I debug with `strace`, I can not see any info about trying to load library (eg. open the .so in /lib or $LD_LIBRARY_PATH).

It seems `plugin` handles my function improperly.
```
$ strace ./test-plugin 2>&1 | grep open
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libdl.so.2", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libpthread.so.0", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
```

### What did you expect to see?
```
panic: dlopen got null

goroutine 1 [running]:
main.main()
        /mnt/e/coding/go/src/github.com/Sora233/test-plugin/main.go:21 +0x108
```

### What did you see instead?

 ```
./test-plugin: symbol lookup error: ./test-plugin: undefined symbol: say_hello
```

**Comment by ianlancetaylor** [comment]
This happens because when dynamically linking with a plugin we force all symbols to be resolved when the program starts (by passing `-znow` to the C linker).  This is necessary because otherwise the symbol resolution may occur when the program is running with a small stack, which will cause it to crash.  This causes your program to fail because there is no definition of `say_hello`.  I can get the same error to occur if I build your program, without importing the "plugin" package, using

    go build -ldflags=-extldflags=-Wl,-znow

Closing because I don't think there is anything we can change here.

**Comment by Sora233** [comment]
Hi @ianlancetaylor , I tried running some codes with`-znow` removed manually and I didn't notice any crash.

Could you explain more about when and how the program crashes?

And if the program may crash during the symbol resolution, does it mean the program will work well after all the symbol resolution finishes?

**Comment by ianlancetaylor** [comment]
Your program doesn't define `say_hello`.  So when the program is linked with `-znow`, the dynamic linker tries to resolve `say_hello`, fails, and exits before starting the program.

If you don't use `-znow`, the dynamic linker does not try to resolve `say_hello` at program startup time.  Instead, it waits until there is a call.  When there is a call, it turns out that at that point `say_hello` is defined, because of the `dlopen` with `RTLD_GLOBAL`, so the program succeeds.

So basically `-znow` can lead to an incorrect failure to start the program if the program happens to use `RTLD_GLOBAL.
